gnc/sub8_perception/nodes/bins_2d.py

```diff
@@ -0,0 +1,173 @@
+        self.bin_type = None
+        self.last_image = None
+        self.last_draw_image = None
+        self.last_image_time = None
+        self.camera_model = None
+        self.pose_service = rospy.Service('vision/bin/2D', VisionRequest2D, self.request_bin)
+        self.image_sub = sub8_ros_tools.Image_Subscriber('/down_camera/image_rect_color', self.image_cb)
+        self.image_pub = sub8_ros_tools.Image_Publisher('/vision/bin_2d/target_info')
+
+        # Occasional status publisher
+        self.timer = rospy.Timer(rospy.Duration(1.0), self.publish_target_info)
+
+        self.bins = {
+            'orange': '/color/bin/orange',
+            'norange': '/color/bin/norange',
+        }
+
+    def request_bin(self, srv):
+        self.bin_type = srv.target_name
+        if (self.last_image != None):
+            print 'requesting', srv
+            response = self.find_single_bin(np.copy(self.last_image), srv.target_name)
+
+            if response is False:
+                print 'did not find'
+                resp = VisionRequest2DResponse(
+                    header=sub8_ros_tools.make_header(frame='/down'),
+                    found=False
+                )
+
+            else:
+                # Fill in
+                center, radius = response
+                resp = VisionRequest2DResponse(
+                    header=Header(stamp=self.last_image_time, frame_id='/down'),
+                    pose=Pose2D(
+                        x=center[0],
+                        y=center[1],
+                        theta=radius
+                    ),
+                    max_x=self.last_image.shape[0],
+                    max_y=self.last_image.shape[1],
+                    camera_info=self.image_sub.camera_info,
+                    found=True
+                )
+            return resp
+
+    def publish_target_info(self, *args):
+        if self.last_image is None:
+            return
+        self.find_bins(np.copy(self.last_image), self.bin_type)
+        if self.last_draw_image is not None:
+            self.image_pub.publish(self.last_draw_image)
+
+    def image_cb(self, image):
+        '''Hang on to last image'''
+        self.last_image = image
+        self.last_image_time = self.image_sub.last_image_time
+        if self.camera_model is None:
+            if self.image_sub.camera_info is None:
+                return
+
+            self.camera_model = image_geometry.PinholeCameraModel()
+            self.camera_model.fromCameraInfo(self.image_sub.camera_info)
+
+    def find_single_bin(self, img, bin_type):
+        """Find the bins and their orientations."""
+        assert bin_type in self.bins[bin_type], "Bins_2d does not know bin color: {}".format(bin_type)
+        if img is not None:
+
+            debug_image = np.copy(img)
+            img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+            zeros = np.array([0])
+            mask = cv2.inRange(img, 127, 255)
+            img = cv2.bitwise_or(mask, zeros)
```

Comment:
"X or False" reduces to "X". You might get better mileage out of .astype(np.bool) or

`img = mask > 0`